src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -305,125 +312,122 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
+                if (TypeIsASetInterface(type))
+                {
+                    if (!bindingPoint.IsReadOnly || bindingPoint.Value is not null)
                     {
-                        if (!bindingPoint.IsReadOnly || bindingPoint.Value is not null)
+                        object? newValue = BindSet(type, (IEnumerable?)bindingPoint.Value, config, options);
+                        if (!bindingPoint.IsReadOnly && newValue != null)
                         {
-                            object? newValue = BindDictionaryInterface(bindingPoint.Value, type, config, options);
-                            if (!bindingPoint.IsReadOnly && newValue != null)
-                            {
-                                bindingPoint.SetValue(newValue);
-                            }
+                            bindingPoint.SetValue(newValue);
                         }
-
-                        return;
                     }
 
-                    // If we don't have an instance, try to create one
-                    if (bindingPoint.Value is null)
+                    return;
+                }
+
+                // -----------------------------------------------------------------------------------------------------------------------------
+                //                         |  bindingPoint |  bindingPoint |
+                //       Interface         |     Value     |   IsReadOnly  |  Behavior
+                // -----------------------------------------------------------------------------------------------------------------------------
+                //  IDictionary<T>         |   not null    |  true/false   | Use the Value instance to populate the configuration
+                //  IDictionary<T>         |     null      |     false     | Create Dictionary<T> instance to populate the configuration
+                //  IDictionary<T>         |     null      |     true      | nothing
+                //  IReadOnlyDictionary<T> | null/not null |     false     | Create Dictionary<K,V> instance, copy over existing values, and populate the configuration
+                //  IReadOnlyDictionary<T> | null/not null |     true      | nothing
+                // -----------------------------------------------------------------------------------------------------------------------------
+                if (TypeIsADictionaryInterface(type))
+                {
+                    if (!bindingPoint.IsReadOnly || bindingPoint.Value is not null)
                     {
-                        // if the binding point doesn't let us set a new instance, there's nothing more we can do
-                        if (bindingPoint.IsReadOnly)
+                        object? newValue = BindDictionaryInterface(bindingPoint.Value, type, config, options);
+                        if (!bindingPoint.IsReadOnly && newValue != null)
```

Comment:
Nit/question (not blocking): Would `is not null` be preferred here? `is (not) null` is used through most of this code, but there are some `!= null`s mixed in.

Reply:
Per https://github.com/dotnet/runtime/pull/91359#discussion_r1312214668, we could def clean up aspects of the reflection code later.